### PR TITLE
 Configuration does not serialize default values

### LINKED ISSUE #1246: Establish a pattern for default configuration values, so we don't have to serialize empty dictionaries or arrays
Configuration files are getting bigger, so we may need to establish a pattern for default configuration values, so we don't have to serialize empty dictionaries or arrays.

For example check [this](https://github.com/Azure/autorest.csharp/blob/feature/v3/samples/AppConfiguration/Generated/Configuration.json) configuration file with a bunch of empty dictionaries and arrays.

### PR DESCRIPTION
**Pull request by chamons**
Today we unconditionally write out all default values, which makes the config files very bloated as we add more and more options.

This change adds infrastructure to know what is default, and not write them out (or read them).

If we ever change the default value, this could bite us however.

- Fixes #1246
- Does not change management configuration values for now

# Checklist

To ensure a quick review and merge, please ensure:
- [x] The PR has a understandable title and description explaining the _why_ and _what_.
- [x] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [x] The branch is recent enough to not have merge conflicts upon creation.

### CODE REVIEW

**Comment by chamons** [comment]
The tl-dr diff is shorter: https://gist.github.com/chamons/6a580fbbd97e5c62708c7ad2f1418885

File: src/AutoRest.CSharp/Common/AutoRest/Plugins/Configuration.cs [hunk 1]
```diff
@@ -49,17 +49,45 @@ public static Configuration GetConfiguration(IPluginCommunication autoRest)
...
+                Get(autoRest, "public-clients"),
+                Get(autoRest, "model-namespace"),
+                Get(autoRest, "head-as-boolean"),
+                Get(autoRest, "skip-csproj-packagereference"),
+                Get(autoRest, "low-level-client"),
                 MgmtConfiguration.GetConfiguration(autoRest)
             );
         }
 
+        private static bool Get (IPluginCommunication autoRest, string option)
```
Comment (AlexanderSher) [inline src/AutoRest.CSharp/Common/AutoRest/Plugins/Configuration.cs:63]: Can we call it "GetBool" or something similar?

Comment (AlexanderSher) [reply src/AutoRest.CSharp/Common/AutoRest/Plugins/Configuration.cs:63]: Also, you can have default value as a third parameter, that way you won't need separate `GetDefault`

File: src/AutoRest.CSharp/Common/AutoRest/Plugins/Configuration.cs [hunk 2]
```diff
@@ -49,17 +49,45 @@ public static Configuration GetConfiguration(IPluginCommunication autoRest)
...
+        {
+            return autoRest.GetValue<bool?>(option).GetAwaiter().GetResult() ?? GetDefault(option)!.Value;
+        }
+
+        public static bool? GetDefault (string option)
```
Comment (pakrym) [inline src/AutoRest.CSharp/Common/AutoRest/Plugins/Configuration.cs:68]: nit: space in method name

**Comment by pakrym** [comment]
Should we do something for mgmt collections?

**Review by pakrym — APPROVED** [review]

**Comment by chamons** [comment]
> Should we do something for mgmt collections?

I reached out to @m-nash and he said there was already work being done (done?) by @allenjzhang in this space.

**Comment by chamons** [comment]
Strange...

```
Error: Unable to start AutoRest Core from C:\Users\VssAdministrator\.autorest\@autorest_core@3.4.1\node_modules\@autorest\core
Error: Unable to start AutoRest Core from C:\Users\VssAdministrator\.autorest\@autorest_core@3.4.1\node_modules\@autorest\core
    at main (D:\a\1\s\autorest.csharp\node_modules\autorest\dist\app.js:222:19)
Invoke : Command failed to execute: npx --no-install autorest --require=D:\a\1\s\autorest.csharp\readme.md --try-require=D:\a\1\s\autorest.csharp\test\TestServerProjects\additionalProperties\readme.md --input-file=D:\a\1\s\autorest.csharp\node_modules\@microsoft.azure\autorest.testserver\swagger\additionalProperties.json   --skip-upgrade-check  --namespace=additionalProperties --output-folder=D:\a\1\s\autorest.csharp\test\TestServerProjects\additionalProperties\Generated
```

But it works locally. Rekicking.

**Pull request merged**